src/Database/LSMTree/Internal/Chunk.hs

```diff
@@ -0,0 +1,98 @@
+module Database.LSMTree.Internal.Chunk
+(
+    -- * Chunks
+    Chunk (Chunk),
+
+    -- * Balers
+    Baler,
+    createBaler,
+    feedBaler,
+    readBalerRemnant
+)
+where
+
+import           Prelude hiding (concat, length, splitAt)
+
+import           Control.Monad.ST.Strict (ST)
+import           Data.STRef.Strict (STRef, newSTRef, readSTRef, writeSTRef)
+import           Data.Vector.Primitive (Vector, concat, length, slice, splitAt)
+import           Data.Word (Word8)
+
+-- * Chunks
+
+-- | A chunk of bytes, typically output during incremental index serialisation.
+newtype Chunk = Chunk (Vector Word8)
+
+-- * Balers
+
+-- Mutable state of a baler.
+data BalerState = BalerState
+                      ![Vector Word8] -- Reverse list of queued blocks
+                      !Int            -- Total size of queued blocks
+
+{-|
+    An object that receives blocks of bytes and repackages them into chunks of
+    preferably a given maximum size.
+-}
+data Baler s = Baler
+                   !Int                  -- Maximum chunk size
+                   !(STRef s BalerState) -- Reference to the mutable state
```

Comment:
Seeing as the maximum chunk size is already known, you could pre-allocate a mutable vector of the appropriate size and copy the blocks into it. I think that might be a nicer alternative to the block queue

Reply:
I don’t think that this alternative would be nicer. It would certainly use a more imperative programming style. Furthermore, it would involve a data structure (the chunk buffer) that contains a part (the not yet written part of the buffer) that doesn’t represent any of the actual data, so that there would be a clear mismatch between the data structure and the data that it represents. All in all, this approach looks quite low-level to me.

This alternative might be preferable if it has clear benefits in terms of memory or time usage, but I even have doubts that it does. With both your and my approach the data has to be copied once, and with your approach the baler always has the whole buffer allocated, while with my approach it can often share data with other parts of the software (when a baler is used for an ordinary index, the serialized keys are shared).

As a result, I’m not really convinced that changing the implementation in the above-mentioned way is what we should do. What do you think?

Reply:
> The intent of the `lsm-tree` package and its sub-components is to be fast, so it is a reasonable trade-off to write lower-level code in favour of performance. I understand that nicer code is probably what we'd all prefer to write, but we can use unsafe functions/features and mutable data structures without sacrificing "nicety" by ensuring safety at the API boundary of this module. Moreover, if we have to sacrifice generality in favour of performance, for example by making `Baler` more specific to the `lsm-tree` use case, then I also think that is a road we should take

A thing to consider here is that the `Chunk` module is primarily intended to support the ordinary index, which should be a general-purpose solution, unlike the compact index, which is to be tailored towards the specific use case of the Cardano blockchain.

Reply:
> > Furthermore, it would involve a data structure (the chunk buffer) that contains a part (the not yet written part of the buffer) that doesn’t represent any of the actual data, so that there would be a clear mismatch between the data structure and the data that it represents.
> 
> I'm not sure what you mean with this mismatch. They are the same bytes, aren't they?

The unwritten part of the chunk buffer would contain bytes that do not correspond to any bytes that are scheduled to be part of the next chunk; so it would contain redundant information. However, I’ve realized meanwhile that the queue data structure also contains redundant information: it remembers where the boundaries between the input blocks are. As a result, the redundancy argument isn’t a valid one.

Reply:
> I think there are benefits: […]

These arguments make sense to me for the most part; so I think I’ll go for the solution with the chunk buffer.

Reply:
Actually, when changing the implementation to use a chunk buffer, I’ll already change it also according to #303.

Reply:
The new implementation is available in 5713127cfc0b9b15fde5f5d04c4ab8edde696dd3.